kernel/objects/rapl-driver-intel/objects/RaplDriverIntel.cc

```diff
@@ -0,0 +1,254 @@
+/* -*- mode:C++; indent-tabs-mode:nil; -*- */
+/* MIT License -- MyThOS: The Many-Threads Operating System
+ *
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use, copy,
+ * modify, merge, publish, distribute, sublicense, and/or sell copies
+ * of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * Copyright 2020 Philipp Gypser and contributors, BTU Cottbus-Senftenberg
+ */
+
+#include "objects/RaplDriverIntel.hh"
+#include "mythos/InvocationBuf.hh"
+#include "util/assert.hh"
+#include "boot/mlog.hh"
+#include "mythos/protocol/RaplDriverIntel.hh"
+#include "mythos/HostInfoTable.hh"
+#include "objects/TypedCap.hh"
+#include "cpu/ctrlregs.hh"
+#include "objects/IntelRegs.hh"
+//#include "objects/IFrame.hh"
+//#include <cmath>
+
+
+namespace mythos {
+
+  extern PhysPtr<HostInfoTable> hostInfoPtrPhys SYMBOL("_host_info_ptr");
+
+  RaplDriverIntel::RaplDriverIntel(){
+    MLOG_INFO(mlog::boot, "RAPL driver startet");
+    // check it is an Intel cpu
+    auto r = x86::cpuid(0);
+    char str[] = "0123456789ab";
+    *reinterpret_cast<uint32_t*>(&str[0]) = r.ebx;
+    *reinterpret_cast<uint32_t*>(&str[4]) = r.edx;
+    *reinterpret_cast<uint32_t*>(&str[8]) = r.ecx;
+    
+    isIntel = true;
+    char intelStr[] = INTEL_STR;
+    for(int i = 0; i < 12; i++){
+      if(str[i] != intelStr[i]){
+        isIntel = false;
+        break;
+      }
+    }
```

Comment:
```
char intel[] = "GenuineIntel";
auto intel32 = reinterpret_cast<uint32_t*>(intel);
auto isIntel = intel32[0] == r.ebx && intel32[1] == r.edx && intel32[3] == r.ecx;
```
However, this probably is undefined behaviour due to the C++ type aliasing rules. Anything can be casted into a char array in order to inspect the object's contents. But the reverse direction is not allowed.

```
uint32_t str[] = {r.ebx, r.edx, r.ecx};
auto isIntel = memcmp(str, "GenuineIntel", 3*sizeof(uint32_t));
```

Reply:
done (implemented memcmp in utils/mstring.hh)